scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala

```diff
@@ -1080,13 +1081,19 @@ class FormatOps(val tree: Tree, val initStyle: ScalafmtConfig) {
 
   def getSpaceAndNewlineAfterCurlyLambda(
       newlines: Int
-  )(implicit style: ScalafmtConfig): (Boolean, NewlineT) =
+  )(implicit style: ScalafmtConfig): (Boolean, NewlineT) = {
+    val tryToSquashLambdaInOneLine =
+      style.newlines.sourceIs(Newlines.fold) || newlines == 0
     style.newlines.afterCurlyLambda match {
-      case NewlineCurlyLambda.never => (true, Newline)
+      case NewlineCurlyLambda.never => (tryToSquashLambdaInOneLine, Newline)
       case NewlineCurlyLambda.always => (false, Newline2x)
       case NewlineCurlyLambda.preserve =>
-        (newlines == 0, if (newlines >= 2) Newline2x else Newline)
+        (
+          tryToSquashLambdaInOneLine,
```

Comment:
this redefines the meaning of `preserve` and might end squashing when the user had a newline (or a couple!).